url0e7d57473a_kyorohiro_HetimaUtil_tgz-pJ8-net_hetimatan_net_http_HttpServer3xxJ8/src/net/hetimatan/util/bitfield/BitField.java

```diff
@@ -0,0 +1,286 @@
+		if(len>inb.lengthPerByte()) {
+			len = inb.lengthPerByte();
+		}
+		for(int i=0;i<out.lengthPerByte();i++) {
+			out.mBitfield[i] = (byte)(0xFF&out.mBitfield[i]);
+		}
+		for(int i=0;i<len;i++) {
+			out.mBitfield[i] = (byte)(0xFF&ina.mBitfield[i]&(~inb.mBitfield[i]));
+		}
+		out.update();
+		return out;
+	}
+
+	public static int numOfOn(BitField field) {
+		int ret = 0;
+		int size = field.lengthPerBit();
+		for(int i=0;i<size;i++) {
+			
+		}
+		return ret;
+	}
+
+	public void update(){}
+	public BitField(int bitsize) {
+		mR = new Random(System.currentTimeMillis());
+		mBitsize = bitsize;
+		int byteSize = bitsize/8;
+		if((bitsize%8)!=0) {
+			byteSize+=1;
+		}
+		mBitfield = new byte[byteSize];
+		oneClear();
+	}
+
+	//
+	// TODO next work following method is wrong
+	//
+	public int getOffPieceAtRandom() {
+		return getPieceAtRandom(true);
+	}
+
+	public int getOnPieceAtRandom() {
+		return getPieceAtRandom(false);
+	}
+
+	//
+	// TODO next work following method is wrong
+	//
+	public int getPieceAtRandom(boolean isOff) {
+		int byteLength  = lengthPerByte();
+		if(byteLength<=0) {
+			return -1;
+		}
+		int ia = mR.nextInt(byteLength);
+		boolean findedAtIA = false;
+		for(int i=ia;i<byteLength;i++) {
+			if(isOff) {
+				if(!isAllOnPerByte(i)) {
+					ia = i;
+					findedAtIA = true;break;
+				}
+			} else {
+				if(!isAllOffPerByte(i)) {
+					ia = i;
+					findedAtIA = true;break;
+				}
+			}
+		}
+
+		if(!findedAtIA) {
+			for(int i=ia;i>=0;i--) {
+				if(isOff) {
+					if(!isAllOnPerByte(i)) {
+						ia = i;
+						findedAtIA = true;break;
+					}
+				} else {
+					if(!isAllOffPerByte(i)) {
+						ia = i;
+						findedAtIA = true;break;
+					}					
+				}
+			}
+		}
+		if(!findedAtIA) {
+			return -1;
+		}
+
+		if(isOff) {
+			return getOffPieceAtRandomPerByte(ia);
+		} else {
+			return getOnPieceAtRandomPerByte(ia);			
+		}
+	}
+
+	protected void shuffle(int[] shufflelist) {
+		int tmp1 = 0;
+		int tmp2 = 0;
+		for(int i=0;i<8;i++) {
+			tmp1 = mR.nextInt(8);
+			tmp2 = shufflelist[i];
+			shufflelist[i] = shufflelist[tmp1];
+			shufflelist[tmp1] = tmp2;
+		}
+	}
+
+	public int getOffPieceAtRandomPerByte(int numPerByte) {
+		return getPieceAtRandomPerByte(numPerByte, true);
+	}
+
+	public int getOnPieceAtRandomPerByte(int numPerByte) {
+		return getPieceAtRandomPerByte(numPerByte, false);
+	}
+
+	public int getPieceAtRandomPerByte(int numPerByte, boolean isOff) {
+		int byteLength  = lengthPerByte();
+		if(byteLength<=0) {
+			return -1;
+		}
+		shuffle(mShuffleList);
+
+		int rn = 8;
+		if(rn>(lengthPerBit()-numPerByte*8)){
+			rn =(lengthPerBit()-numPerByte*8); 
+		}
+		for(int i=0;i<8;i++) {
+			if((numPerByte*8+mShuffleList[i])<lengthPerBit()&&
+					isOff != isOn(numPerByte*8+mShuffleList[i])) {
+				return (numPerByte*8+mShuffleList[i]);
+			}			
+		}
+		return -1;
+	}
+
+	public boolean isAllOff() {
+		int len =lengthPerBit();
+		for(int i=0;i<len;i++) {
+			if(isOn(i)) {
+				return false;
+			}
+		}
+		return true;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 158:164, 168:174.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.